compiler/src/iree/compiler/Codegen/Common/ReshapePatterns.cpp

```diff
@@ -968,6 +991,105 @@ struct FoldInnerBitcastIntoInterfaceTensorStore
   }
 };
 
+/// Similar to FoldInnerBitcastIntoInterfaceTensorLoad, but handles the
+/// bufferized load instead:
+///   hal.interface.binding.subspan -> amdgpu.fat_raw_buffer_cast ->
+///   iree_codegen.load_from_buffer -> iree_tensor_ext.bitcast
+struct FoldInnerBitcastIntoLoadFromBuffer
+    : OpRewritePattern<IREE::TensorExt::BitCastOp> {
+  using Base::Base;
+
+  LogicalResult matchAndRewrite(IREE::TensorExt::BitCastOp bitcastOp,
+                                PatternRewriter &rewriter) const override {
+    Value bitcastSrc = bitcastOp.getSource();
+
+    // Step 1: Check for iree_codegen.load_from_buffer.
+    auto loadOp = bitcastSrc.getDefiningOp<IREE::Codegen::LoadFromBufferOp>();
+    if (!loadOp) {
+      return failure();
+    }
+    Value loadBuffer = loadOp.getBuffer();
+
+    // Step 2: Check for amdgpu.fat_raw_buffer_cast.
+    auto bufferCastOp = loadBuffer.getDefiningOp<amdgpu::FatRawBufferCastOp>();
+    if (!bufferCastOp) {
+      return failure();
+    }
+    Value bufferCastSource = bufferCastOp.getSource();
+
+    // Step 3: Check for hal.interface.binding.subspan.
+    auto subspanOp =
+        bufferCastSource.getDefiningOp<IREE::HAL::InterfaceBindingSubspanOp>();
+    if (!subspanOp) {
+      return failure();
+    }
+
+    // Now we have the full bufferized load chain verified.
+    auto subspanType = cast<MemRefType>(subspanOp.getType());
+    auto bitcastSrcType = cast<RankedTensorType>(bitcastSrc.getType());
+    auto bitcastResType = cast<RankedTensorType>(bitcastOp.getType());
+
+    // Verify that only the inner most dimension is changed by the bitcast by
+    // comparing dynamic and static sizes for equality.
+    if (bitcastOp.getSourceDims() != bitcastOp.getResultDims() ||
+        bitcastSrcType.getShape().drop_back() !=
+            bitcastResType.getShape().drop_back() ||
+        ShapedType::isDynamic(bitcastSrcType.getShape().back())) {
+      return failure();
+    }
+
+    // Fail if the inner most dim doesn't match between subspan and tensor.
+    if (subspanType.getShape().back() != bitcastSrcType.getShape().back()) {
+      return failure();
+    }
+
+    int64_t newInnerSize = bitcastResType.getShape().back();
+    int64_t oldInnerSize = bitcastSrcType.getShape().back();
+
+    // Adjust the innermost dimension size.
+    MemRefLayoutAttrInterface newSubspanLayout;
+    SmallVector<int64_t> newSubspanShape(subspanType.getShape());
+    newSubspanShape.back() = newInnerSize;
+    // Scale offset and strides by (newInnerSize / oldInnerSize).
+    if (auto stridedLayout =
+            dyn_cast_or_null<StridedLayoutAttr>(subspanType.getLayout())) {
```

Comment:
nit: we prefer `dyn_cast_if_present`.

`dyn_cast_or_null` has been marked that it will be deprecated for three years. I just did a cleanup for IREE today: https://github.com/iree-org/iree/commit/33db935a33ddab6a8358418c0ae6dc263e56b0bb